packages/services/storage/src/index.ts

```diff
@@ -2511,6 +2548,104 @@ export async function createStorage(connection: string, maximumPoolSize: number)
           "id" = ${args.oidcIntegrationId}
       `);
     },
+
+    idMutex: {
+      async lock(id, { signal }) {
+        return Promise.race([
+          new Promise<never>((_, reject) => {
+            const listener = () => {
+              signal.removeEventListener('abort', listener);
+              reject(new Error('Locking aborted'));
+            };
+            signal.addEventListener('abort', listener);
+          }),
+          (async () => {
+            const lockConn = await getLockConn();
+            if (signal.aborted) {
+              throw new Error('Locking aborted');
+            }
+
+            // postgres advisory locks uses bigints as lock keys,
+            // we therefore hash the provided lock id and use the hash
+            const idInt = hashFnv32a(id);
+
+            // wait if there's a lock in the process
+            let lock;
+            while ((lock = locks.get(id))) {
+              await lock.p;
+              if (signal.aborted) {
+                throw new Error('Locking aborted');
+              }
+            }
+
+            // only one lock can be acquired per process
+            let release!: () => void;
+            const p = new Promise<void>(resolve => (release = resolve));
+            locks.set(id, {
+              p,
+              release,
+            });
+
+            // wait and acquire lock on the database (intra-process sync)
+            let advisoryLock = false;
+            try {
+              while (!advisoryLock) {
+                ({ advisoryLock } = await lockConn.one<{ advisoryLock: boolean }>(
+                  sql`select pg_try_advisory_lock(${idInt}) as advisoryLock`,
+                ));
+                if (!advisoryLock) {
+                  // only sleep if not locked so that the loop can resolve fast if locked
+                  await new Promise(resolve => setTimeout(resolve, 1_000));
+                }
+                if (signal.aborted) {
+                  throw new Error('Locking aborted');
+                }
+              }
+            } catch (err) {
+              // lock not acquired in database, unlock and bubble error
+              locks.delete(id);
+              release();
+              if (advisoryLock) {
+                lockConn.query(sql`select pg_advisory_unlock(${idInt})`).catch(err => {
+                  // TODO: what to do instead?
+                  console.error(
```

Comment:
I think we can use logger.warn for now - if we notice it keeps happening a lot we can think more about it.